### ISSUE TITLE
How to specify a restrictive topic in PubSub IoT policy?

### ISSUE THREAD

**Issue by abhi-at-bitsy** [aws-amplify/amplify-js#9408]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

PubSub

### Amplify Categories

Not applicable

### Environment information

<details>

```
# Put output below this line


```

</details>

### Describe the bug

I followed the steps mentioned [here](https://docs.amplify.aws/lib/pubsub/getting-started/q/platform/js/#step-1-create-iam-policies-for-aws-iot) to create an IoT policy and attached that to the logged in user identity, and it works just fine.

But the policy created is widely open and I want it to be topic specific. Example a user belonging to firmA should be able to only subscribe to topics which starts with firmA/* (firmA/templates, firmA/users, firmA/advisors/templates and so on)
 
I tried multiple policies but nothing is working. Few examples which didn't work:

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "iot:*"
      ],
      "Resource": [
        "arn:aws:iot:us-east-2:{accountNumber}:topic/firmA",
        "arn:aws:iot:us-east-2:{accountNumber}:topic/firmA/*",
        "arn:aws:iot:us-east-2:{accountNumber}:topicfilter/firmA"
        "arn:aws:iot:us-east-2:{accountNumber}:topicfilter/firmA/*"
      ]
    }
  ]
}
```

### Expected behavior

```
const sub = PubSub.subscribe('firmA/users').subscribe({
      next: (data) => console.log('Message received', data),
      error: (error) => console.error(error),
      complete: () => console.log('Done'),
    });
```

should not throw an error on the browser. It works if I attach a policy which is just a wild card on resource.

### Reproduction steps

I followed the steps mentioned [here](https://docs.amplify.aws/lib/pubsub/getting-started/q/platform/js/#step-1-create-iam-policies-for-aws-iot) to create an IoT policy and attached that to the logged in user identity, and it works just fine.

But the policy created is widely open and I want it to be topic specific. Example a user belonging to firmA should be able to only subscribe to topics which starts with firmA/* (firmA/templates, firmA/users, firmA/advisors/templates and so on)
 
I tried multiple policies but nothing is working. Few examples which didn't work:

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "iot:*"
      ],
      "Resource": [
        "arn:aws:iot:us-east-2:{accountNumber}:topic/firmA",
        "arn:aws:iot:us-east-2:{accountNumber}:topic/firmA/*",
        "arn:aws:iot:us-east-2:{accountNumber}:topicfilter/firmA"
        "arn:aws:iot:us-east-2:{accountNumber}:topicfilter/firmA/*"
      ]
    }
  ]
}
```

**Comment by ashika01** [comment]
@abhi-at-bitsy here is a doc on how to set the policy in iot.

https://docs.aws.amazon.com/iot/latest/developerguide/pub-sub-policy.html

you might have to make sure both data and config have restrictive access. I will try this in a bit and update here

**Comment by stocaaro** [comment]
Hello @abhi-at-bitsy,

I took some time this morning to reproduce your results and review the instructions @ashika01 linked.

With the policy above, I see the error `"AMQJS0008I Socket closed."`.

I was able to work around this by adding a broader Connect permission on top of the topic filtering.

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "iot:Connect",
      "Resource": "arn:aws:iot:us-east-1:{accountNumber}:*"
    },
    {
      "Effect": "Allow",
      "Action": "iot:*",
      "Resource": [
        "arn:aws:iot:us-east-1:{accountNumber}:topic/firmA/*",
        "arn:aws:iot:us-east-1:{accountNumber}:topicfilter/firmA/*"
      ]
    }
  ]
}
```

WIth this, I setup a second policy with `*` permissions and two users, one with the limited policy and the other with the `*` policy.

With this setup I was able to verify that both users could Publish and Receive from the topic `firmA/users` where the limited user couldn't Publish and Receive against `firmB/users`.

Is this consistent with the outcome you are looking for to cover your usecase?
